pyanaconda/display.py

```diff
@@ -210,41 +207,67 @@ def do_startup_x11_actions():
     # pylint: disable=environment-modify
     os.environ['XDG_CONFIG_DIRS'] = xdg_config_dirs
 
-    def x11_preexec():
+    os.environ["XDG_SESSION_TYPE"] = "wayland"
+
+    def wl_preexec():
         # to set GUI subprocess SIGINT handler
         signal.signal(signal.SIGINT, signal.SIG_IGN)
 
-    childproc = util.startProgram(["gnome-kiosk", "--display", ":1", "--sm-disable", "--x11"],
-                                  env_add={'XDG_DATA_DIRS': xdg_data_dirs},
-                                  preexec_fn=x11_preexec)
-    WatchProcesses.watch_process(childproc, "gnome-kiosk")
+    # lets compile arguments for the run-in-new-session script
+    argv = ["run-in-new-session.py",
+            "--user", "root",
+            "--service", "anaconda",
+            "--session-type", "wayland",
+            "--session-class", "user"]
 
+    if headless:
+        # headless (remote connection) - stay on VT1 where connection info is
+        argv.extend(["--vt", "1"])
+    else:
+        # local display - switch to VT6 & show GUI there
+        argv.extend(["--vt", "6"])
 
-def set_x_resolution(runres):
-    """Set X server screen resolution.
+    # add the generic GNOME Kiosk invocation
+    argv.extend(["gnome-kiosk", "--sm-disable",
+                 "--wayland", "--no-x11",
+                 "--wayland-display", constants.WAYLAND_SOCKET_NAME])
 
-    :param str runres: a resolution specification string
-    """
-    try:
-        log.info("Setting the screen resolution to: %s.", runres)
-        util.execWithRedirect("xrandr", ["-d", ":1", "-s", runres])
-    except RuntimeError:
-        log.error("The X resolution was not set")
-        util.execWithRedirect("xrandr", ["-d", ":1", "-q"])
+    # remote access needs gnome-kiosk to start in headless mode &
+    # configure a virtual monitor
+    if headless:
+        # check virtual monitor  resolution has been set
+        if headless_resolution is None:
+            # use default value
+            headless_resolution = "1280x1024"
+        argv.extend(["--headless", "--virtual-monitor", headless_resolution])
```

Comment:
Running the headless g-r-d should automatically create its own virtual monitor, with a resolution that matches the RDP clients.

Reply:
I seem to remember it would not work without us setting the resolution with this option, but I'll re-check.

Reply:
BTW, if I drop the `--virtual-monitor` option in the headless mode, it does not seem to be happy at all:
![no_virtual_monitor](https://github.com/rhinstaller/anaconda/assets/829558/44115c03-cd04-4b9e-a85b-4f223cdeb801)

All this before it becomes possible to connect - I guess it default to 0x0 pixels resolution or something ?

The Anaconda GUI needs to be running (and might even do things by itself as IIRC it should be possible to connect to an automated kickstart installation running in GUI mode) before the first client connects, so some sensible default "headless" screen size should be used, to be possibly overridden by the clients connecting.

IIRC @jstodola mentioned that the GUI already does handle screen resolution changes in some VM installation use cases.

Still, I wonder what would happen if a second user connects (which seems to be possible with RDP) & requests a different screen resolution ?

Reply:
With GNOME 45 / F39, I think you need to tweak the gsettings (`gsettings set org.gnome.desktop.remote-desktop.rdp screen-share-mode extend`), but in GNOME 46 / F40 it'll be automatic.

Reply:
There might be bugs in clients that doesn't handle the case where they try to show up before the RDP client connects, though (the warnings seem to be about that).